README.md
# graylog-plugin-misp

[![License](https://img.shields.io/badge/license-SSPL-green)](https://www.mongodb.com/licensing/server-side-public-license)

## License

This plugin is released under version 1 of the [Server Side Public License (SSPL)](LICENSE).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## [0.1.0](https://github.com/airbus-cyber/graylog-plugin-misp/tree/0.1.0)
* First release


validation/test.py
# to create and populate the test venv:
# * python3 -m venv venv
# * source venv/bin/activate
# * pip install -r requirements.txt 
# to execute these tests:
# * activate venv
#   source ./venv/bin/activate
# * execute tests
#   python -m unittest

from unittest import TestCase
from graylog_server import GraylogServer


class Test(TestCase):

    def setUp(self) -> None:
        self._graylog = GraylogServer('../runtime')
        self._graylog.start()

    def tearDown(self) -> None:
        self._graylog.stop()

    def test_start_should_load_plugin(self):
        logs = self._graylog.extract_logs()
        self.assertIn('INFO : org.graylog2.bootstrap.CmdLineTool - Loaded plugin: MISP', logs)

    def test_pipeline_rule_functions_should_include_misp_lookup(self):
        self.assertTrue(self._graylog.has_pipeline_rule_function('misp_lookup'))

    def test_pipeline_rule_function_misp_lookup_should_accept_a_value_argument_as_its_first_parameter(self):
        misp_lookup = self._graylog.get_pipeline_rule_function('misp_lookup')
        first_parameter = misp_lookup['params'][0]
        self.assertEqual(first_parameter['name'], 'value')

    def test_create_pipeline_rule_with_function_misp_lookup_should_not_fail(self):
        response = self._graylog.create_pipeline_rule()
        self.assertEqual(response.status_code, 200)

    def test_pipeline_with_misp_lookup_should_add_a_new_field_to_message(self):
        self._graylog.create_pipeline_rule()
        self._graylog.create_pipeline()

        with self._graylog.create_gelf_input() as gelf_inputs:
            gelf_inputs.send({'src_addr': '192.168.1.254'})

            last_message = self._graylog.wait_for_message()
            self.assertFalse(last_message['found_in_misp'])



validation/graylog_rest_api.py
import requests
import time
from urllib import parse
from requests.exceptions import ConnectionError
from graylog_inputs import GraylogInputs

STREAM_ALL_MESSAGES = "000000000000000000000001"
_AUTH = ('admin', 'admin')
_HEADERS = {"X-Requested-By": "test-program"}


class GraylogRestApi:

    def _build_url(self, path):
        return parse.urljoin('http://127.0.0.1:9000/api/', path)

    def get(self, path):
        url = self._build_url(path)
        response = requests.get(url, auth=_AUTH, headers=_HEADERS)
        print('GET {} => {}'.format(url, response.status_code))
        return response

    def put(self, path, payload):
        url = self._build_url(path)
        response = requests.put(url, json=payload, auth=_AUTH, headers=_HEADERS)
        print('PUT {} {} => {}'.format(url, payload, response.status_code))

    def post(self, path, payload=None):
        url = self._build_url(path)
        response = requests.post(url, json=payload, auth=_AUTH, headers=_HEADERS)
        print('POST {} {} => {}'.format(url, payload, response.status_code))
        return response

    def _input_is_running(self, identifier):
        response = self.get('system/inputstates/')
        body = response.json()
        for state in body['states']:
            if state['id'] != identifier:
                continue
            return state['state'] == 'RUNNING'
        return False

    def create_gelf_input(self):
        payload = {
            'configuration': {
                'bind_address': '0.0.0.0',
                'decompress_size_limit': 8388608,
                'max_message_size': 2097152,
                'number_worker_threads': 8,
                'override_source': None,
                'port': 12201,
                'recv_buffer_size': 1048576,
                'tcp_keepalive': False,
                'tls_cert_file': '',
                'tls_client_auth': 'disabled',
                'tls_client_auth_cert_file': '',
                'tls_enable': False,
                'tls_key_file': 'admin',
                'tls_key_password': 'admin',
                'use_null_delimiter': True
            },
            'global': True,
            'title': 'Inputs',
            'type': 'org.graylog2.inputs.gelf.tcp.GELFTCPInput'
        }
        response = self.post('system/inputs', payload)
        identifier = response.json()['id']
        while not self._input_is_running(identifier):
            time.sleep(.1)
        return GraylogInputs()

    def wait_until_graylog_has_started(self):
        """
        We wait until the default deflector is up, as it seems to be the last operation done on startup
        This might have to change in the future, if graylog changes its ways...
        :return:
        """
        print('Waiting for graylog to start...')

        # TODO move as a method in _graylog_rest_api
        #only for 60s maximum
        while True:
            try:
                response = self.get('system/deflector')
                body = response.json()
                if body['is_up']:
                    break
            except ConnectionError:
                pass
            time.sleep(1)

    def create_stream_with_rule(self, title, field, value):
        response = self.get('system/indices/index_sets')
        default_index_set_identifier = response.json()['index_sets'][0]['id']
        stream = {
            'description': title,
            'index_set_id': default_index_set_identifier,
            'remove_matches_from_default_stream': False,
            'title': title
        }
        response = self.post('streams', stream)
        stream_identifier = response.json()['stream_id']
        rule = {
            'description': '',
            'field': field,
            'inverted': False,
            'type': 1,
            'value': value
        }
        self.post('streams/{}/rules'.format(stream_identifier), rule)
        self.post('streams/{}/resume'.format(stream_identifier))
        return stream_identifier



src/main/java/com/airbus_cyber_security/graylog/MispPlugin.java
/*
 * Copyright (C) 2018 Airbus CyberSecurity (SAS)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package com.airbus_cyber_security.graylog;

import com.google.auto.service.AutoService;
import org.graylog2.plugin.Plugin;
import org.graylog2.plugin.PluginMetaData;
import org.graylog2.plugin.PluginModule;

import java.util.Arrays;
import java.util.Collection;

/**
 * Implement the Plugin interface here.
 */
@AutoService(Plugin.class)
public class MispPlugin implements Plugin {
    @Override
    public PluginMetaData metadata() {
        return new MispMetaData();
    }

    @Override
    public Collection<PluginModule> modules () {
        return Arrays.<PluginModule>asList();
    }
}

src/main/java/com/airbus_cyber_security/graylog/MispMetaData.java
/*
 * Copyright (C) 2018 Airbus CyberSecurity (SAS)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package com.airbus_cyber_security.graylog;

import org.graylog2.plugin.PluginMetaData;
import org.graylog2.plugin.ServerStatus;
import org.graylog2.plugin.Version;

import java.net.URI;
import java.util.Collections;
import java.util.Set;

/**
 * Implement the PluginMetaData interface here.
 */
public class MispMetaData implements PluginMetaData {
    private static final String PLUGIN_PROPERTIES = "com.airbus-cyber-security.graylog.graylog-plugin-misp/graylog-plugin.properties";

    @Override
    public String getUniqueId() {
        return "com.airbus-cyber-security.graylog.MispPlugin";
    }

    @Override
    public String getName() {
        return "MISP";
    }

    @Override
    public String getAuthor() {
        return "Airbus CyberSecurity";
    }

    @Override
    public URI getURL() {
        return URI.create("https://www.airbus-cyber-security.com");
    }

    @Override
    public Version getVersion() {
        return Version.fromPluginProperties(getClass(), PLUGIN_PROPERTIES, "version", Version.from(0, 0, 1));
    }

    @Override
    public String getDescription() {
        return "MISP database lookup functions for the Graylog Pipeline Processor";
    }

    @Override
    public Version getRequiredVersion() {
        return Version.fromPluginProperties(getClass(), PLUGIN_PROPERTIES, "graylog.version", Version.from(4, 2, 0));
    }

    @Override
    public Set<ServerStatus.Capability> getRequiredCapabilities() {
        return Collections.emptySet();
    }
}

src/main/java/com/airbus_cyber_security/graylog/pipelineprocessor/functions/MispModule.java
/*
 * Copyright (C) 2018 Airbus CyberSecurity (SAS)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package com.airbus_cyber_security.graylog.pipelineprocessor.functions;

import com.google.inject.Binder;
import com.google.inject.multibindings.MapBinder;
import com.google.inject.TypeLiteral;
import org.graylog2.plugin.PluginModule;
import org.graylog.plugins.pipelineprocessor.ast.functions.Function;

public class MispModule extends PluginModule {

    @Override
    protected void configure() {
        addMessageProcessorFunction(MispLookupFunction.NAME, MispLookupFunction.class);
    }

    private void addMessageProcessorFunction(String name, Class<? extends Function<?>> functionClass) {
        MapBinder<String, Function<?>> processorFunctionBinder = createProcessorFunctionBinder();
        processorFunctionBinder.addBinding(name).to(functionClass).asEagerSingleton();
    }

    private MapBinder<String, Function<?>> createProcessorFunctionBinder() {
        Binder binder = binder();
        TypeLiteral<String> keyType = TypeLiteral.get(String.class);
        TypeLiteral<Function<?>> valueType = new TypeLiteral<Function<?>>() {};
        return MapBinder.newMapBinder(binder, keyType, valueType);
    }
}


src/main/java/com/airbus_cyber_security/graylog/pipelineprocessor/functions/MispLookupFunction.java
/*
 * Copyright (C) 2018 Airbus CyberSecurity (SAS)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package com.airbus_cyber_security.graylog.pipelineprocessor.functions;

import org.graylog.plugins.pipelineprocessor.EvaluationContext;
import org.graylog.plugins.pipelineprocessor.ast.functions.AbstractFunction;
import org.graylog.plugins.pipelineprocessor.ast.functions.ParameterDescriptor;
import org.graylog.plugins.pipelineprocessor.ast.functions.FunctionArgs;
import org.graylog.plugins.pipelineprocessor.ast.functions.FunctionDescriptor;

public class MispLookupFunction extends AbstractFunction<Boolean> {

    public static final String NAME = "misp_lookup";

    private final ParameterDescriptor<String, String> valueParameter = ParameterDescriptor.string("value").description("Value of the attribute to look for in MISP.").build();
    private final ParameterDescriptor<String, String> attributeTypeParameter = ParameterDescriptor.string("attribute_type").description("The type of attribute to look for in MISP.").build();

    @Override
    public Boolean evaluate(FunctionArgs functionArgs, EvaluationContext evaluationContext) {
        return false;
    }

    @Override
    public FunctionDescriptor<Boolean> descriptor() {
        return FunctionDescriptor.<Boolean>builder()
                .name(NAME)
                .description("Searches for the presence of an attribute with given type in MISP.")
                .params(valueParameter, attributeTypeParameter)
                .returnType(Boolean.class)
                .build();
    }
}


validation/graylog_server.py
import subprocess
import time
from requests.exceptions import ConnectionError
from graylog_rest_api import GraylogRestApi
from graylog_inputs import GraylogInputs


class GraylogServer:

    def __init__(self, docker_compose_path):
        self._docker_compose_path = docker_compose_path
        self._rest_api = GraylogRestApi()

    def start(self):
        subprocess.run(['docker-compose', 'up', '--detach'], cwd=self._docker_compose_path)
        self.wait_until_graylog_has_started()

    def extract_logs(self, line_count='all'):
        tail_option = '--tail={}'.format(line_count)
        return subprocess.check_output(['docker-compose', 'logs', tail_option, '--no-color', 'graylog'], cwd=self._docker_compose_path, universal_newlines=True)

    def stop(self):
        subprocess.run(['docker-compose', 'down'], cwd=self._docker_compose_path)

    def has_pipeline_rule_function(self, name):
        return self.get_pipeline_rule_function(name) is not None

    def get_pipeline_rule_function(self, name):
        pipeline_rule_functions = self._rest_api.get('system/pipelines/rule/functions')
        for pipeline_rule_function in pipeline_rule_functions.json():
            if pipeline_rule_function['name'] == name:
                return pipeline_rule_function
        return None

    def _input_is_running(self, identifier):
        response = self._rest_api.get('system/inputstates/')
        body = response.json()
        for state in body['states']:
            if state['id'] != identifier:
                continue
            return state['state'] == 'RUNNING'
        return False

    def create_gelf_input(self):
        payload = {
            'configuration': {
                'bind_address': '0.0.0.0',
                'decompress_size_limit': 8388608,
                'max_message_size': 2097152,
                'number_worker_threads': 8,
                'override_source': None,
                'port': 12201,
                'recv_buffer_size': 1048576,
                'tcp_keepalive': False,
                'tls_cert_file': '',
                'tls_client_auth': 'disabled',
                'tls_client_auth_cert_file': '',
                'tls_enable': False,
                'tls_key_file': 'admin',
                'tls_key_password': 'admin',
                'use_null_delimiter': True
            },
            'global': True,
            'title': 'Inputs',
            'type': 'org.graylog2.inputs.gelf.tcp.GELFTCPInput'
        }
        response = self._rest_api.post('system/inputs', payload)
        identifier = response.json()['id']
        while not self._input_is_running(identifier):
            time.sleep(.1)
        return GraylogInputs()

    def wait_until_graylog_has_started(self):
        """
        We wait until the default deflector is up, as it seems to be the last operation done on startup
        This might have to change in the future, if graylog changes its ways...
        :return:
        """
        print('Waiting for graylog to start...')

        # TODO move as a method in _graylog_rest_api
        #only for 60s maximum
        while True:
            try:
                response = self._rest_api.get('system/deflector')
                body = response.json()
                if body['is_up']:
                    break
            except ConnectionError:
                pass
            time.sleep(1)

    def create_stream_with_rule(self, title, field, value):
        response = self._rest_api.get('system/indices/index_sets')
        default_index_set_identifier = response.json()['index_sets'][0]['id']
        stream = {
            'description': title,
            'index_set_id': default_index_set_identifier,
            'remove_matches_from_default_stream': False,
            'title': title
        }
        response = self._rest_api.post('streams', stream)
        stream_identifier = response.json()['stream_id']
        rule = {
            'description': '',
            'field': field,
            'inverted': False,
            'type': 1,
            'value': value
        }
        self._rest_api.post('streams/{}/rules'.format(stream_identifier), rule)
        self._rest_api.post('streams/{}/resume'.format(stream_identifier))
        return stream_identifier

    def create_pipeline_rule(self):
        payload = {
            'description': '',
            'source': 'rule "MISP Lookup: src_addr"\nwhen\n has_field("src_addr")\nthen\n let src_addr = to_string($message.src_addr);\n let found = misp_lookup(src_addr, "ip-src");\n set_field("found_in_misp", found);\nend\n'
        }
        return self._rest_api.post('system/pipelines/rule', payload)

    def create_pipeline(self):
        payload = {
            'description': '',
            'source': 'pipeline "pipeline"\nstage 0 match either\nrule "MISP Lookup: src_addr"\nend',
        }
        response = self._rest_api.post('system/pipelines/pipeline', payload)
        pipeline = response.json()
        payload = {
            'pipeline_id': pipeline['id'],
            'stream_ids': ['000000000000000000000001']
        }
        self._rest_api.post('system/pipelines/connections/to_pipeline', payload)

    def _get_messages(self):
        payload = {
            'queries': [{
                'search_types': [{
                    'type': 'messages'
                }]
            }]
        }
        response = self._rest_api.post('views/search/sync', payload)
        body = response.json()
        for result in body['results'].values():
            for search_type in result['search_types'].values():
                messages = search_type['messages']
                return messages

    def wait_for_message(self):
        while True:
            messages = self._get_messages()
            if len(messages) > 0:
                message = messages[0]
                return message['message']
            time.sleep(1)
